test-app/tests/integration/environment-test.js

```diff
@@ -0,0 +1,364 @@
+      },
+    },
+  });
+
+  exclaimTest('value reference resolution', {
+    env: { foo: 'bar' },
+    resolveFieldMeta: (path) => path,
+    ui: {
+      value: { key: { $bind: 'foo' } },
+      async $validate(assert) {
+        const value = this.config.value;
+        assert.strictEqual(get(value, 'key'), 'bar');
+        assert.strictEqual(this.env.metaForField(value, 'key'), 'foo');
+
+        set(value, 'key', 'ok');
+        assert.strictEqual(get(this.env, 'foo'), 'ok');
+      },
+    },
+  });
+
+  exclaimTest('chained reference resolution', {
+    env: { foo: 'bar' },
+    resolveFieldMeta: (path) => path,
+    ui: {
+      value: { key: { child: { $bind: 'foo' } } },
+      async $validate(assert) {
+        const value = this.config.value;
+        const subvalue = get(value, 'key');
+        assert.strictEqual(get(subvalue, 'child'), 'bar');
+
+        assert.strictEqual(this.env.metaForField(value, 'key'), undefined);
+        assert.strictEqual(this.env.metaForField(value, 'key.child'), 'foo');
+        assert.strictEqual(this.env.metaForField(subvalue, 'child'), 'foo');
+
+        set(subvalue, 'child', 321);
+        assert.strictEqual(get(value, 'key.child'), 321);
+        assert.strictEqual(get(this.env, 'foo'), 321);
+
+        set(this.env, 'foo', 999);
+        assert.strictEqual(get(subvalue, 'child'), 999);
+      },
+    },
+  });
+
+  exclaimTest('array values', {
+    env: { array: A([1, 2, 3]) },
+    ui: {
+      ref: { array: { $bind: 'array' } },
+      async $validate(assert) {
+        const ref = this.config.ref;
+        assert.deepEqual(get(this.env, 'array'), [1, 2, 3]);
+        assert.deepEqual(get(ref, 'array'), [1, 2, 3]);
+
+        const envArray = get(this.env, 'array');
+        const refArray = get(ref, 'array');
+        assert.strictEqual(envArray[1], 2);
+        assert.strictEqual(refArray[1], 2);
+
+        envArray.replace(1, 1, [100]);
+        assert.deepEqual(envArray, [1, 100, 3]);
+        assert.deepEqual(refArray, [1, 100, 3]);
+
+        envArray.pushObject(4);
+        assert.deepEqual(envArray, [1, 100, 3, 4]);
+        assert.deepEqual(refArray, [1, 100, 3, 4]);
+
+        refArray.shiftObject();
+        assert.deepEqual(envArray, [100, 3, 4]);
+        assert.deepEqual(refArray, [100, 3, 4]);
+      },
+    },
+  });
+
+  exclaimTest('environment extension', {
+    env: { a: 'bar' },
+    ui: {
+      async $validate(assert) {
+        const base = this.env;
+        const ext1 = base.extend({ b: 'qux' });
+        const ext2 = base.extend({ a: 'newbar' });
+
+        assert.strictEqual(get(base, 'a'), 'bar');
+        assert.strictEqual(get(base, 'b'), undefined);
+
+        assert.strictEqual(get(ext1, 'a'), 'bar');
+        assert.strictEqual(get(ext1, 'b'), 'qux');
+
+        assert.strictEqual(get(ext2, 'a'), 'newbar');
+        assert.strictEqual(get(ext2, 'b'), undefined);
+
+        // Writing to an inherited key updates at the level it was inherited from
```

Comment:
These are interesting. Can you explain when/why the environment is extended like this?

Reply:
I see now - it's done in components that render nested exclaim configuration.